File: neetcode150/medium/153-Find-Minimum-in-Rotated-Sorted-Array.py
# Solution 1
class Solution:
  def findMin(self, nums):
    begin = 0
    end = len(nums) - 1
    res = 100
        
    while begin <= end:
      mid = (begin + end) // 2
            
      if nums[mid] <= nums[begin] and nums[mid] <= nums[end]:
        res = min(res, nums[mid])
            
      if nums[begin] >= nums[mid] and nums[end] <= nums[mid]:
        begin = mid + 1
      elif nums[end] <= nums[mid]: 
        begin = mid + 1
      elif nums[begin] < nums[mid]:
        end = mid - 1
      elif nums[end] > nums[mid]:
        end = mid - 1
    return res

# Solution 2
class Solution:
  def findMin(self, nums):
    l = 0
    r = len(nums) - 1
    while l < r:
      mid = (l + r) // 2

      if nums[mid] > nums[r]:
        l = mid + 1
      else:
        r = mid
    return nums[l]

File: neetcode150/medium/153-Find-Minimum-in-Rotated-Sorted-Array.cpp
#include <vector>

using namespace std;

/**
 * \c Solution
 * 
 * 1. Use the top number of the array to determine how it's rotated
 * 2. Adjust i and j indices accordingly
 * 
 * Time  : O(log n)
 * Space : O(1)
 */
class Solution {
public:
    int findMin(vector<int>& nums) {
        int i = 0;
        int j = nums.size() - 1;
        while (i < j) {
            int mid = i + (j - i) / 2;
            if (nums[mid] < nums[j]) {
                j = mid;
            } else {
                i = mid + 1;
            }
        }
        return nums[i];
    }
};